File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/UnassignedVariablesWalker.cs
using System.Collections.Generic;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class UnassignedVariablesWalker : DataFlowPass
	{
		private readonly HashSet<Symbol> _result;

		protected override bool SuppressRedimOperandRvalueOnPreserve => false;

		protected override bool IgnoreOutSemantics => false;

		protected override bool EnableBreakingFlowAnalysisFeatures => true;

		private UnassignedVariablesWalker(FlowAnalysisInfo info)
			: base(info, suppressConstExpressionsSupport: false, trackStructsWithIntrinsicTypedFields: true)
		{
			_result = new HashSet<Symbol>();
		}

		internal static HashSet<Symbol> Analyze(FlowAnalysisInfo info)
		{
			UnassignedVariablesWalker unassignedVariablesWalker = new UnassignedVariablesWalker(info);
			try
			{
				return unassignedVariablesWalker.Analyze() ? unassignedVariablesWalker._result : new HashSet<Symbol>();
			}
			finally
			{
				unassignedVariablesWalker.Free();
			}
		}

		protected override void ReportUnassigned(Symbol local, SyntaxNode node, ReadWriteContext rwContext, int slot = -1, BoundFieldAccess boundFieldAccess = null)
		{
			if (local.Kind == SymbolKind.Field)
			{
				Symbol nodeSymbol = GetNodeSymbol(boundFieldAccess);
				if ((object)nodeSymbol != null)
				{
					_result.Add(nodeSymbol);
				}
			}
			else
			{
				_result.Add(local);
			}
			base.ReportUnassigned(local, node, rwContext, slot, boundFieldAccess);
		}

		internal override void AssignLocalOnDeclaration(LocalSymbol local, BoundLocalDeclaration node)
		{
			if (!local.IsStatic)
			{
				base.AssignLocalOnDeclaration(local, node);
			}
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/UnassignedVariablesWalker.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System
Imports System.Collections.Generic

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class UnassignedVariablesWalker
		Inherits DataFlowPass
		Private ReadOnly _result As HashSet(Of Microsoft.CodeAnalysis.VisualBasic.Symbol)

		Protected Overrides ReadOnly Property EnableBreakingFlowAnalysisFeatures As Boolean
			Get
				Return True
			End Get
		End Property

		Protected Overrides ReadOnly Property IgnoreOutSemantics As Boolean
			Get
				Return False
			End Get
		End Property

		Protected Overrides ReadOnly Property SuppressRedimOperandRvalueOnPreserve As Boolean
			Get
				Return False
			End Get
		End Property

		Private Sub New(ByVal info As FlowAnalysisInfo)
			MyBase.New(info, False, True)
			Me._result = New HashSet(Of Microsoft.CodeAnalysis.VisualBasic.Symbol)()
		End Sub

		Friend Shared Function Analyze(ByVal info As FlowAnalysisInfo) As HashSet(Of Microsoft.CodeAnalysis.VisualBasic.Symbol)
			Dim symbols As HashSet(Of Microsoft.CodeAnalysis.VisualBasic.Symbol)
			Dim unassignedVariablesWalker As Microsoft.CodeAnalysis.VisualBasic.UnassignedVariablesWalker = New Microsoft.CodeAnalysis.VisualBasic.UnassignedVariablesWalker(info)
			Try
				symbols = If(unassignedVariablesWalker.Analyze(), unassignedVariablesWalker._result, New HashSet(Of Microsoft.CodeAnalysis.VisualBasic.Symbol)())
			Finally
				unassignedVariablesWalker.Free()
			End Try
			Return symbols
		End Function

		Friend Overrides Sub AssignLocalOnDeclaration(ByVal local As LocalSymbol, ByVal node As BoundLocalDeclaration)
			If (Not local.IsStatic) Then
				MyBase.AssignLocalOnDeclaration(local, node)
			End If
		End Sub

		Protected Overrides Sub ReportUnassigned(ByVal local As Microsoft.CodeAnalysis.VisualBasic.Symbol, ByVal node As SyntaxNode, ByVal rwContext As AbstractFlowPass(Of DataFlowPass.LocalState).ReadWriteContext, Optional ByVal slot As Integer = -1, Optional ByVal boundFieldAccess As Microsoft.CodeAnalysis.VisualBasic.BoundFieldAccess = Nothing)
			If (local.Kind <> SymbolKind.Field) Then
				Me._result.Add(local)
			Else
				Dim nodeSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbol = MyBase.GetNodeSymbol(boundFieldAccess)
				If (nodeSymbol IsNot Nothing) Then
					Me._result.Add(nodeSymbol)
				End If
			End If
			MyBase.ReportUnassigned(local, node, rwContext, slot, boundFieldAccess)
		End Sub
	End Class
End Namespace